ArjunaCore/arjuna/tests/classes/com/hp/mwtests/ts/arjuna/recovery/RecoverySuspendTest.java

```diff
@@ -0,0 +1,159 @@
+public class RecoverySuspendTest {
+
+    private static RecoveryManager _manager;
+    private static RecordTypeMap _recordTypeMap;
+
+    // fields read with byteman
+    private final static int periodicRecoveryPeriod = 5;
+    private final static int recoveryBackoffPeriod = 1;
+
+    @BeforeAll
+    public static void beforeClass() {
+
+        RecoveryEnvironmentBean recoveryConfig = recoveryPropertyManager.getRecoveryEnvironmentBean();
+        String[] modules = {
+                // the module to test
+                "com.arjuna.ats.internal.arjuna.recovery.AtomicActionRecoveryModule"
+        };
+
+        // Let's go quick on this
+        recoveryConfig.setRecoveryBackoffPeriod(recoveryBackoffPeriod);
+        recoveryConfig.setPeriodicRecoveryPeriod(periodicRecoveryPeriod);
+        // don't sign off until the store is empty
+        recoveryConfig.setWaitForFinalRecovery(true);
+
+        // the test set of modules
+        recoveryConfig.setRecoveryModuleClassNames(Arrays.asList(modules));
+
+        // obtain a new RecoveryManager with the above config:
+        _manager = RecoveryManager.manager(RecoveryManager.INDIRECT_MANAGEMENT);
+        // recovery can start
+        _manager.startRecoveryManagerThread();
+
+        _recordTypeMap = new RecordTypeMap() {
+            @SuppressWarnings("unchecked")
+            public Class getRecordClass ()
+            {
+                return BytemanControlledRecord.class;
+            }
+
+            public int getType ()
+            {
+                return RecordType.USER_DEF_FIRST0;
+            }
+        };
+
+        RecordTypeManager.manager().add(_recordTypeMap);
+    }
+
+    @BeforeEach
+    public void beforeEach() {
+        // In case the Recovery Manager was suspended, this will resume it
+        _manager.resume();
+    }
+
+    @AfterAll
+    public static void afterClass() {
+        RecordTypeManager.manager().remove(_recordTypeMap);
+    }
+
+    @AfterEach
+    public void afterEach() {
+        if (!TxControl.isEnabled()) {
+            // Re-activate transaction creation
+            TxControl.enable();
+        }
+    }
+
+    @Test
+    public void testSuspensionWheneThereArentTxnsToRecover() {
+
+        // The Transaction System needs to be disabled, i.e. no new transactions can be created
+        TxControl.disable();
+
+        // Makes sure that the transaction reaper completed all transactions
+        TransactionReaper.transactionReaper().waitForAllTxnsToTerminate();
+
+        long startSuspension = System.currentTimeMillis();
+
+        // Synchronization is not needed (i.e. async = true) as isWaitForFinalRecovery() == true
+        _manager.suspend(true);
+        long durationOfSuspension = System.currentTimeMillis() - startSuspension;
+
+        Assertions.assertTrue(durationOfSuspension <
+                // In milliseconds
+                1000L * (periodicRecoveryPeriod + recoveryBackoffPeriod), String.valueOf(durationOfSuspension));
```

Comment:
Can this be done in some other way than a timed test? The timing is not predictable. Can't you just assert that https://github.com/jmfinelli/narayana/pull/4/files#diff-6438c48ab6e59780ef50133a412623a12fe5dcb9a0df1912a37c5f38542f2ac4R242-R244 has been called and it has not returned true (carefully worded this because I think you can assume it was called due to the algorithm and given the test we know it should not have returned true even if it was ran more than once)?

Reply:
What you envisioned might be a bit convoluted. As an alternative, I modified this test to use BytemanControlledRecord with an extra method to artificially set the `greenFlag`. Do you think this is enough to test the case without recovery?

Reply:
I modified this test forcing the `greenFlag` to true through `BytemanControlledResource`. As you already asked somewhere else to add an explanation for the new method, I suppose we can solve this conversation, can't we?

Reply:
I think this thread can be marked as resolved, thank you

Reply:
Thanks Tom